pallet/account-migration/src/lib.rs

```diff
@@ -16,174 +16,166 @@
 
-			<T as pallet::Config>::Currency::deposit_creating(&new_pub_key.into(), amount.into());
-			Self::deposit_event(Event::Claim { old_pub_key, new_pub_key, amount });
+			Self::deposit_event(Event::Migrated { from, to });
 
 			Ok(())
 		}
 	}
 	#[pallet::validate_unsigned]
-	impl<T: Config> ValidateUnsigned for Pallet<T>
-	where
-		AccountIdOf<T>: From<H160>,
-		BalanceOf<T>: From<Balance>,
-	{
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
 		type Call = Call<T>;
 
 		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
-			let Call::claim_to { chain_id, old_pub_key, new_pub_key, sig } = call else {
+			// The migration destination was already taken by someone.
+			const E_ACCOUNT_ALREADY_EXISTED: u8 = 0;
+			// The migration source was not exist.
+			const E_ACCOUNT_NOT_FOUND: u8 = 1;
+			// Invalid signature.
+			const E_INVALID_SIGNATURE: u8 = 2;
+
+			let Call::migrate { from, to, signature } = call else {
 				return InvalidTransaction::Call.into();
 			};
 
-			if *chain_id != <T as pallet_evm::Config>::ChainId::get() {
-				return InvalidTransaction::BadProof.into();
+			if !<Accounts<T>>::contains_key(from) {
+				return InvalidTransaction::Custom(E_ACCOUNT_NOT_FOUND).into();
 			}
-			if !Balances::<T>::contains_key(old_pub_key) {
-				return InvalidTransaction::BadSigner.into();
+			if <frame_system::Account<T>>::contains_key(to) {
+				return InvalidTransaction::Custom(E_ACCOUNT_ALREADY_EXISTED).into();
 			}
 
-			let message = ClaimMessage::new(
-				<T as pallet_evm::Config>::ChainId::get(),
-				old_pub_key,
-				new_pub_key,
-			);
-			if let Ok(signer) = Public::from_slice(old_pub_key.as_ref()) {
-				let is_valid = sig.verify(&blake2_256(&message.raw_bytes())[..], &signer);
-
-				if is_valid {
-					return ValidTransaction::with_tag_prefix("MigrateClaim")
-						.priority(TransactionPriority::max_value())
-						.propagate(true)
-						.build();
-				}
+			let message = sr25519_signable_message(T::Version::get().spec_name.as_ref(), to);
+
+			if verify_sr25519_signature(from, &message, signature) {
+				ValidTransaction::with_tag_prefix("account-migration")
+					.and_provides(from)
+					.priority(100)
+					.longevity(TransactionLongevity::max_value())
+					.propagate(true)
+					.build()
+			} else {
+				InvalidTransaction::Custom(E_INVALID_SIGNATURE).into()
 			}
-			InvalidTransaction::BadSigner.into()
 		}
 	}
 }
+pub use pallet::*;
 
-/// ClaimMessage is the metadata that needs to be signed when the user invokes claim dispatch.
-///
-/// It consists of three parts, namely the chain_id, the AccountId32 account for darwinia 1.0, and
-/// the H160 account for darwinia 2.0.
-pub struct ClaimMessage<'m> {
-	pub chain_id: u64,
-	pub old_pub_key: &'m AccountId32,
-	pub new_pub_key: &'m H160,
+fn sr25519_signable_message(spec_name: &[u8], account_id_20: &AccountId20) -> Message {
+	hashing::blake2_256(&[spec_name, b"::account-migration", &account_id_20.0].concat())
```

Comment:
Please confirm this spec. Then we could sync with @isunaslabs.

Reply:
LGTM, maybe add more magic words here is better to avoid duplicate replay sig with other cases.
Such as
```
b"::account-migration_for_darwinia_2_0_from_1_0"
```

Reply:
We could also follow this https://github.com/darwinia-network/darwinia-common/blob/6a9392cfb9fe2c99b1c2b47d0c36125d61991bb7/frame/bridge/ecdsa-authority/src/primitives.rs#L42

```
hash(hash(chain id + spec name + pallet) + data)
```

Reply:
> We could also follow this https://github.com/darwinia-network/darwinia-common/blob/6a9392cfb9fe2c99b1c2b47d0c36125d61991bb7/frame/bridge/ecdsa-authority/src/primitives.rs#L42
> 
> 
> 
> ```
> 
> hash(hash(chain id + spec name + pallet) + ethereum address)
> 
> ```

OK, can also improve it later.